game-core/src/main/java/games/strategy/triplea/util/MovableUnitsFilter.java

```diff
@@ -0,0 +1,205 @@
+package games.strategy.triplea.util;
+
+import com.google.common.base.Preconditions;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.MoveDescription;
+import games.strategy.engine.data.PlayerId;
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitCollection;
+import games.strategy.triplea.delegate.AbstractMoveDelegate;
+import games.strategy.triplea.delegate.AbstractMoveDelegate.MoveType;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.MoveValidator;
+import games.strategy.triplea.delegate.UndoableMove;
+import games.strategy.triplea.delegate.UnitComparator;
+import games.strategy.triplea.delegate.data.MoveValidationResult;
+import games.strategy.triplea.delegate.data.MustMoveWithDetails;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Utility class for filtering a unit list to the subset of those units that can move on a route.
+ */
+public final class MovableUnitsFilter {
+  /** The result of the filter operation. */
+  @Getter
+  public static class Result {
+    // The filtered units and dependents that can move on the route.
+    private final List<Unit> unitsWithDependents;
+    // An error message, if no units could move on the route.
+    @Nullable private String errorMessage;
```

Comment:
Would this potentially be more clear named "noUnitsCanMoveMessage"?
Similar for the warning, "someUnitsCanMoveMessage"

Reply:
Maybe. Another option is to just have message and an enum of (all units can move, some units can move, no units can move) and a message.

Reply:
Looking at other parts of code, it looks like having an enum be the result would lead to a better separation of concerns.

For example, MovePanel already 'owns' this status string:
```
      setStatusWarningMessage("Not all units can move there");
```

It's a bit odd to have some status strings owned by a filter class, and others by the panel. We can avoid the nullity SRP issues as well by having a single result flag, and just let the MovePanel own the status/error/warning strings.

WDYT @asvitkine ?

Reply:
Went with the enum, but still keeping the message as coming from MoveValidator.